language-plutus-core/src/Language/PlutusCore/Instance/Eq/Internal.hs

```diff
@@ -0,0 +1,208 @@
+-- | An internal module that defines functions for deciding equality of values of core data types.
+
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+module Language.PlutusCore.Instance.Eq.Internal
+    ( LR (..)
+    , RL (..)
+    , EqRename
+    , ScopedEqRename
+    , runEqRename
+    , runScopedEqRename
+    , withRenamedTwins
+    , eqNameM
+    , eqTypeM
+    , eqTermM
+    , eqProgramM
+    ) where
+
+import           PlutusPrelude
+
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Rename.Monad
+import           Language.PlutusCore.Type
+
+import           Control.Lens
+
+-- | From left to right.
+newtype LR a = LR
+    { unLR :: a
+    } deriving (Generic)
+
+-- | From right to left.
+newtype RL a = RL
+    { unRL :: a
+    } deriving (Generic)
+
+-- | A left @a@ and a right @a@.
+data Bilateral a = Bilateral
+    { _bilateralL :: a
+    , _bilateralR :: a
+    }
+
+makeLenses ''Bilateral
+
+instance Wrapped (LR a)
+instance Wrapped (RL a)
+instance HasUnique name unique => HasUnique (LR name) (LR unique)
+instance HasUnique name unique => HasUnique (RL name) (RL unique)
+
+instance Semigroup a => Semigroup (Bilateral a) where
+    Bilateral l1 r1 <> Bilateral l2 r2 = Bilateral (l1 <> l2) (r1 <> r2)
+
+instance Monoid a => Monoid (Bilateral a) where
+    mempty = Bilateral mempty mempty
+
+-- To rename from left to right is to update the left renaming.
+instance HasRenaming ren unique => HasRenaming (Bilateral ren) (LR unique) where
+    renaming = bilateralL . renaming . coerced @(Renaming unique)
+
+-- To rename from right to left is to update the right renaming.
+instance HasRenaming ren unique => HasRenaming (Bilateral ren) (RL unique) where
+    renaming = bilateralR . renaming . coerced @(Renaming unique)
+
+-- | The type of a runnable equality check. @Maybe ()@ is isomorphic to 'Bool' and we use it
+-- instead of 'Bool', because this unlocks the convenient and readable do-notation and allows for
+-- automatic short-circuiting, which would be tedious with @Rename (Bilateral ren) Bool@.
```

Comment:
Can't decide if this is brilliant or awful :p

Reply:
The former :)

Reply:
Given we might have HIE support soonish, it'd be better to use a newtyped `BoolM` and a few derivings - that way you get the information on hover and can even attach this documentation to it.